url/url.cpp

```diff
@@ -51,4 +101,1341 @@ std::string blob_url_create(Origin const &origin) {
+        if ((r == 10 && !util::is_digit(input[i])) || (r == 16 && !util::is_hex_digit(input[i]))
+                || (r == 8 && !util::is_octal_digit(input[i]))) {
+            return std::nullopt;
+        }
+    }
+
+    return {{std::stoul(std::string{input}, nullptr, r), v_err}};
+}
+
+// https://url.spec.whatwg.org/#concept-ipv6-parser
+std::optional<std::array<std::uint16_t, 8>> UrlParser::parse_ipv6(std::string_view input) {
+    std::array<std::uint16_t, 8> address;
+
+    std::size_t piece_index = 0;
+
+    std::optional<std::size_t> compress;
+
+    std::size_t pointer = 0;
+
+    if (input[pointer] == ':') {
+        if (!input.substr(1).starts_with(":")) {
+            validation_error(ValidationError::IPv6InvalidCompression);
+
+            return std::nullopt;
+        }
+
+        pointer += 2;
+
+        piece_index++;
+
+        compress = piece_index;
+    }
+
+    while (pointer < input.size()) {
+        if (piece_index == 8) {
+            validation_error(ValidationError::IPv6TooManyPieces);
+
+            return std::nullopt;
+        }
+
+        if (input[pointer] == ':') {
+            if (compress.has_value()) {
+                validation_error(ValidationError::IPv6MultipleCompression);
+
+                return std::nullopt;
+            }
+
+            pointer++;
+
+            piece_index++;
+
+            compress = piece_index;
+
+            continue;
+        }
+
+        std::uint64_t value = 0;
+        std::size_t length = 0;
+
+        for (; length < 4 && util::is_hex_digit(input[pointer]); pointer++, length++) {
+            value *= 0x10 + std::stoul(std::string{input[pointer]}, nullptr, 16);
+        }
+
+        if (input[pointer] == '.') {
+            if (length == 0) {
+                validation_error(ValidationError::IPv4InIPv6InvalidCodePoint);
+
+                return std::nullopt;
+            }
+
+            pointer -= length;
+
+            if (piece_index > 6) {
+                validation_error(ValidationError::IPv4InIPv6TooManyPieces);
+
+                return std::nullopt;
+            }
+
+            std::size_t numbers_seen = 0;
+
+            while (pointer < input.size()) {
+                std::optional<std::uint64_t> ipv4_piece;
+
+                if (numbers_seen > 0) {
+                    if (input[pointer] == '.' && numbers_seen < 4) {
+                        pointer++;
+                    } else {
+                        validation_error(ValidationError::IPv4InIPv6InvalidCodePoint);
+
+                        return std::nullopt;
+                    }
+                }
+
+                if (!util::is_digit(input[pointer])) {
+                    validation_error(ValidationError::IPv4InIPv6InvalidCodePoint);
+
+                    return std::nullopt;
+                }
+
+                while (util::is_digit(input[pointer])) {
+                    std::uint64_t number = std::stoul(std::string{input[pointer]}, nullptr, 10);
```

Comment:
Try to use std::from_chars instead. It doesn't require null-terminated strings and lets you check for errors in a sane way.